app/drivers/usb/cdc.c

```diff
@@ -172,11 +178,10 @@ static uint8_t cdc_data_out(USBD_HandleTypeDef *p_dev, uint8_t epnum)
 	if (epnum != CDC_OUT_EP)
 		return HAL_OK;
 
-	self.rx_len = HAL_PCD_EP_GetRxCount(self.p_pcd, epnum);
-
-	status = HAL_PCD_EP_Receive(self.p_pcd, CDC_OUT_EP, self.rx_buff, USB_FS_MAX_PACKET_SIZE);
-
-	xSemaphoreGiveFromISR(self.rx_done_semaphore, &xHigherPriorityTaskWoken);
+	self.rx_buf.len = HAL_PCD_EP_GetRxCount(self.p_pcd, epnum);
+	status = HAL_PCD_EP_Receive(self.p_pcd, CDC_OUT_EP, self.rx_buf.data, USB_FS_MAX_PACKET_SIZE);
+ 
+	xQueueSendFromISR(self.rx_queue_handle, &self.rx_buf, &xHigherPriorityTaskWoken);
```

Comment:
So like I said, the HAL_PCD_EP_Receive sets up the buffer for _next_ transfer, I think we should do the queue send before the HAL_PCD_EP_Receive call.

Reply:
If I move it, it doesn't work well anymore. Not sure exactly how this works tbh.